mbuild/compound.py

```diff
@@ -225,8 +247,171 @@ def particles_by_name(self, name):
+            return True
+        else:
+            return False
+
+    def max_rigid(self):
+        """Return the maximum rigid body ID contained in the Compound.
+
+        This is usually used by compound.root to determine the maximum
+        rigid_id in the system as a whole.
+
+        Returns
+        -------
+        int or None
+            The maximum rigid body ID contained in the Compound. If no
+            rigid body IDs are found, None is returned
+
+        """
+        if self.rigid:
+            if self.children:
+                return max(successor.rigid_id for successor in self.successors() if successor.rigid_id is not None)
+            else:
+                return self.rigid_id
+        else:
+            return None
+
+    def rigid_ids(self):
+        """Generate rigid_ids for all Particles in the Compound.
+
+        Yields
+        ------
+        int
+            The rigid_id of the next Particle in the Compound
+
+        """
+        for particle in self.particles():
+            yield particle.rigid_id
+
+    def rigid_particles(self, rigid_id=None):
+        """Generate all particles in rigid bodies.
+
+        Parameters
+        ----------
+        rigid_id : int, optional
+            Include only particles with this rigid body ID
+
+        """
+        for particle in self.particles():
+            if particle.rigid_id is not None:
+                yield particle
+
+    def set_rigid(self, name=False, rigid_id=None):
+        """Treat the Compound and all successors as a single rigid body.
+
+        Provides a unique rigid body ID to a Compound and its successors.
+        If 'name' is specified, only sucessors with this name are provided
+        with a rigid_id.  This is useful for Compounds where only some of
+        the particles should be rigid (such as a tethered nanoparticle with
+        a rigid core).
+
+        Parameters
+        ----------
+        name : str, optional
+            Include only particles with this name in the rigid body.
+        rigid_id : int, optional
+            Manually set the rigid body ID for this body.
+            Note: This is not recommended, as mBuild keeps track of all
+            rigid body ID's and manually overriding this can lead to 
+            unwanted behavior.
+
+        """
+        if rigid_id is None:
+            if self.root.max_rigid() is not None:
+                rigid_id = self.root.max_rigid() + 1
+            else:
+                rigid_id = 0
+        self.rigid_id = rigid_id
+        for successor in self.successors():
+            if name and successor.name != name:
+                pass
+            else:
+                successor.rigid_id = rigid_id
+        if rigid_id and rigid_id > self.root.max_rigid() + 1:
+            warn("Specified 'rigid_id' > max rigid_id + 1. This will lead to inconsistent rigid body numbering. Consider running reset_rigid().")
+
+    def _set_rigid_only(self, rigid_id):
```

Comment:
This method name doesn't seem to quite convey what it means. If I understand correctly, it sets the rigid id of self and all successors that do not already have a rigid id. Those that do already have one get a bump? Should there be a "force all" mechanism to overwrite?